### PR TITLE
Select final

### LINKED ISSUE #11722: Improve performance of FINAL modifier for Collapsing/Replacing MergeTree
**Use case**
Clickhouse is a very good DB for load-and-analyze type of pattern, but its lack of primary key enforcement constraints limits it for typical monitoring case, when there is a need to query both most-recent (live) and historical data at the same time.
Some implementation is available with Replacing MergeTree allowing to collapse rows with same keys, however there is no 100% assureness on avoiding duplicates when data is flowing live, except using FINAL clause. 
Other descriptions of use case are in #8223,  #5125.  
Parallel FINAL #10463 is not really the solution here, since data may be queried from 3 month ago to today spanning across billions of rows. 

**Describe the solution you'd like**
1. Allow to specify parameter to not-do merge across partitions (merge-tree setting or system setting). This will let CH to assume that each partition already owns unique set of PKs. 
1. Skip merging of partitions when FINAL is specified IF (there is only one active part left in partition AND part.level>0 (has been merged at least once))

This will guarantee that previous days/hours/weeks/month data that has been optimized and merged to single part (by OPTIMIZE FINAL for partition that could be managed by ETL/externally by whatever is loading the data) won't require additional merges, and merge will only be done on most recent partition for which the data is being loaded. 

**Describe alternatives you've considered**
There are many alternatives in other issues #8223,  #5125, but they don't seem to be feasible for analytics DB. 
There is no real way to do it relatively easier in application.

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Do not merge parts across partitions in SELECT FINAL

Detailed description / Documentation draft:
Define a new setting `do_not_merge_across_partitions_select_final`. If it's true we will merge parts only with the same partition, moreover, if there is only one part in partition and it's level > 0 we won't merge it. This resolves https://github.com/ClickHouse/ClickHouse/issues/11722

### CODE REVIEW

**Comment by UnamedRus** [comment]
>  Define a new setting do_not_merge_across_partitions_select_final. If it's true we will merge parts only with the same partition, moreover, if there is only one part in partition and it's level > 0 we won't merge it.

There is a possibility that partition was attached from ./detached dir or from another table and already has level > 0 and duplicates, so it isn't a reliable way to check this, or we would need to reset the level of that partition to zero in that kind of parts manipulation.

**Comment by alexey-milovidov** [comment]
Maybe better to have simple rule: one part in partition - do not postprocess it.

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 1]
```diff
@@ -1237,144 +1237,200 @@ Pipe MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
...
-            /// Connect i-th merge with i-th input port of every copier.
-            for (size_t j = 0; j < ports.size(); ++j)
+        pipe.transform([&](OutputPortRawPtrs ports)
+        {
+            Processors processors;
+            std::vector<OutputPorts::iterator> output_ports;
+            processors.reserve(ports.size() + num_streams);
+            output_ports.reserve(ports.size());
+
+            LOG_DEBUG(log, "Output ports size: {}", ports.size());
```
Comment (CurtizJ) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1401]: Did you forget to remove debug output? If you want to leave it, please, add more context to message.

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 2]
```diff
@@ -1237,144 +1237,200 @@ Pipe MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
...
+            sort_description.emplace_back(header.getPositionByName(sort_columns[i]), 1, 1);
 
-            case MergeTreeData::MergingParams::VersionedCollapsing:
-                return std::make_shared<VersionedCollapsingTransform>(header, pipe.numOutputPorts(),
-                           sort_description, data.merging_params.sign_column, max_block_size);
+        auto get_merging_processor = [&]() -> MergingTransformPtr
+        {
+            switch (data.merging_params.mode)
+            {
+                case MergeTreeData::MergingParams::Ordinary: {
```
Comment (CurtizJ) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1336]: Style. According to out style guide, curly bracket should be on new line or can be ommitted.

**Review by CurtizJ — CHANGES_REQUESTED** [review]
It will be useful to add a perf test and also to check the difference in performance with case, when all partitions are merged.

**Comment by Avogar** [comment]
> It will be useful to add a perf test and also to check the difference in performance with case, when all partitions are merged.

@CurtizJ I've never written perf tests before, can you explain what should I do in performance test? Just make a query with `SELECT ... FINAL`? And how to check the difference in performance in different cases in such tests?

**Comment by CurtizJ** [comment]
You can find examples of performance tests in tests/performance. A test is .xml file with bunch of queries for measuring time of their executing and possible some additional queries for preparation. Also you can set specific settings for tests. A good [example](https://github.com/ClickHouse/ClickHouse/blob/master/tests/performance/read_in_order_many_parts.xml), where all sections exist. It's recommended for execution time of query to be in 0.1-1s.

> And how to check the difference in performance in different cases in such tests?

It's ok to compare two modes of `FINAL` locally, just to be sure, that performance has been improved.

**Comment by Avogar** [comment]
> You can find examples of performance tests in tests/performance. Test is a .xml file with bunch of queries for measuring time of their executing and possible some additional queries for preparation. Also you can set specific settings for tests. A good [example](https://github.com/ClickHouse/ClickHouse/blob/master/tests/performance/read_in_order_many_parts.xml), where all sections exist. It's recommended for execution time of query to be in 0.1-1s.
> 
> Yes, just add a query with `SELECT ... FINAL` and your setting to check that in the future performance of this behaviour won't be broken.
> 
> > And how to check the difference in performance in different cases in such tests?
> 
> It's ok to compare two modes of `FINAL` locally, just to be sure, that performance has been improved.

Thanks for clear explanation! I will try to do it at this week

File: tests/performance/optimized_select_final.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    <create_query>
+        CREATE TABLE optimized_select_final (t DateTime, x Int32)
+        ENGINE = ReplacingMergeTree()
+        PARTITION BY toYYYYMM(t) ORDER BY x
+    </create_query>
+
+    <fill_query>INSERT INTO optimized_select_final SELECT toDate('2000-01-01'), number FROM numbers(1000000)</fill_query>
+
+    <fill_query>INSERT INTO optimized_select_final SELECT toDate('2020-01-01'), number FROM numbers(1000000)</fill_query>
```
Comment (akuzm) [inline tests/performance/optimized_select_final.xml:14]: It makes sense to add `optimize table` as the final fill query -- this way, you can be sure that all background merges have completed before the test queries start, and will get more stable results.

**Comment by Avogar** [comment]
@CurtizJ I see this exception in my perf test: `DB::Exception: Unknown setting do_not_merge_across_partitions_select_final.` But I added this setting in this pr and I can't understand the reason of exception, it works locally, do you know what's wrong?

**Comment by alexey-milovidov** [comment]
@Avogar It because performance test runs two servers: old and new.
This error is Ok for old server.

**Review by CurtizJ — APPROVED** [review]

**Comment by CurtizJ** [comment]
`00840_long_concurrent_select_and_drop_deadlock` is broken in master.

**Comment by den-crane** [comment]
hm, I don't see much improvement:

```sql

CREATE TABLE T (p Int64, d Date, some_col String)
Engine=ReplacingMergeTree 
Partition by toYYYYMM(d)
Order by (p, d);


insert into T select number, '2020-01-01', '' from numbers(100000000);
insert into T select number, '2020-01-01', 'updated' from numbers(100000000);

optimize table T final;

select distinct(_part) from T;
┌─_part──────────┐
│ 202001_1_192_4 │
└────────────────┘

select max(d) from T final where some_col = 'updated';
Elapsed: 2.095 sec. Processed 100.00 million rows, 2.60 GB (47.73 million rows/s., 1.24 GB/s.)

select max(d) from T final where some_col = 'updated' SETTINGS do_not_merge_across_partitions_select_final = 1;
Elapsed: 2.063 sec. Processed 100.00 million rows, 2.60 GB (48.47 million rows/s., 1.26 GB/s.)

expected:
select max(d) from T  where some_col = 'updated';
Elapsed: 0.202 sec. Processed 100.00 million rows, 1.80 GB (494.28 million rows/s., 8.90 GB/s.)

```

**Comment by Avogar** [comment]
> hm, I don't see much improvement:
 Yes, you're right. I forgot to add parallel select in this case. I added it here https://github.com/ClickHouse/ClickHouse/pull/19375

**Pull request merged**